airbyte-integrations/connectors/source-dixa/source_dixa/source.py

```diff
@@ -0,0 +1,128 @@
+# MIT License
+#
+# Copyright (c) 2020 Airbyte
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in all
+# copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+
+
+from abc import ABC
+from datetime import date, datetime, timedelta
+from typing import Any, Iterable, List, Mapping, MutableMapping, Optional, Tuple
+
+import requests
+from airbyte_cdk.sources import AbstractSource
+from airbyte_cdk.sources.streams import Stream
+from airbyte_cdk.sources.streams.http import HttpStream
+from airbyte_cdk.sources.streams.http.auth import TokenAuthenticator
+
+
+DATE_FORMAT = "%Y-%m-%d"
+
+
+class DixaStream(HttpStream, ABC):
+    url_base = "https://exports.dixa.io/v1/"
+    date_format = DATE_FORMAT
+    max_days_in_query = 31  # see https://support.dixa.help/en/articles/174-export-conversations-via-api
+
+    def __init__(self, start_date: date, **kwargs) -> None:
+        super().__init__(**kwargs)
+        self.start_date = start_date
+        self.current_date = datetime.now().date()
+        self.created_after = start_date
+        self.created_before = None
+
+    def next_page_token(self, response: requests.Response) -> Optional[Mapping[str, Any]]:
```

Comment:
This type of date-based partitioning is a perfect use case for [stream slicing](https://docs.airbyte.io/contributing-to-airbyte/python/concepts/stream_slices). It also has the added benefit of getting you 90% of the way to incremental syncs (slices also work with full refresh syncs but I'd recommend going the extra bit to get incremental). 

To implement slicing you'd just define a method on the stream, `stream_slices` which returns all the date intervals you want to query data from (sorted): 

```
def stream_slices() -> Iterable:
 current_start_date = self.start_date
 end_date = datetime.now().date()
 slices = []
 while current_start_date < end_date: 
   slices.append({
      "created_after": current_start_date.strftime(), 
      "created_before": min(current_start_date + timedelta(days=max_days_in_query), end_date).strftime()
  })
  return slices
```

then you can leverage these intervals one by one in request params: 

```
def request_params(self, stream_slice : Mapping [str, Any] = None, **kwargs): 
    return stream_slice
```

at which point you no longer need the `next_page_token` method since the [docs mention there is no pagination for the export endpoint](https://support.dixa.help/en/articles/174-export-conversations-via-api)

To go one extra and implement incremental, you'll need to implement the `get_updated_state` method. You'll also probably want to do `current_start_date = max(self.start_date, stream_state.get("start_date", self.start_date))`.

Reply:
That makes a lot of sense. I started reading about these slices but stopped when I figured they were mostly used for incremental. I'll take another stab at it.

Reply:
And follow up with incremental :-)